atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStore.java

```diff
@@ -0,0 +1,231 @@
+        cassandraKeyValueService.clientPool.runWithRetry(client -> {
+            BoundData boundData = getCurrentBoundData(client);
+            byte[] currentBound = boundData.bound();
+            byte[] currentBackupBound = boundData.backupBound();
+
+            if (isReadableLong(currentBound)) {
+                Preconditions.checkState(currentBackupBound == null || !isReadableLong(currentBackupBound),
+                        "We had both backup and active bounds readable! This is unexpected; please contact support!");
+                log.info("[RESTORE] Didn't restore from backup, because the current timestamp is already readable.");
+                return null;
+            }
+
+            ByteBuffer casQueryBuffer = CassandraTimestampUtils.constructCheckAndSetMultipleQuery(
+                    ImmutableMap.of(
+                            CassandraTimestampUtils.ROW_AND_COLUMN_NAME,
+                            Pair.create(CassandraTimestampUtils.INVALIDATED_VALUE, currentBackupBound),
+                            CassandraTimestampUtils.BACKUP_COLUMN_NAME,
+                            Pair.create(currentBackupBound, PtBytes.EMPTY_BYTE_ARRAY)));
+            executeQueryUnchecked(client, casQueryBuffer);
+            return null;
+        });
+    }
+
+    private void checkOperationWasApplied(Optional<Long> expected, long target, CqlResult result) {
+        if (!CassandraTimestampUtils.wasOperationApplied(result)) {
+            throw constructConcurrentTimestampStoreException(expected, target, result);
+        }
+    }
+
+    private static ConcurrentModificationException constructConcurrentTimestampStoreException(
+            Optional<Long> expected,
+            long target,
+            CqlResult result) {
+        long actual = CassandraTimestampUtils.getLongValueFromApplicationResult(result);
+        String expectedValueString = expected.isPresent() ? expected.get().toString() : "null";
+
+        String msg = "Unable to CAS from {} to {}."
+                + " Timestamp limit changed underneath us (limit in memory: {}, stored in DB: {}).";
+        ConcurrentModificationException except = new ConcurrentModificationException(
+                String.format(replaceBracesWithStringFormatSpecifier(msg),
+                        expectedValueString,
+                        target,
+                        expectedValueString,
+                        actual));
+        log.error(msg, expected, target, expected, actual);
+        DebugLogger.logger.error("Thread dump: {}", ThreadDumps.programmaticThreadDump());
+        throw except;
+    }
+
+    private static String replaceBracesWithStringFormatSpecifier(String msg) {
+        return msg.replaceAll("\\{\\}", "%s");
+    }
+
+    private static CqlResult executeQueryUnchecked(Cassandra.Client client, ByteBuffer queryBuffer) {
+        try {
+            return client.execute_cql3_query(queryBuffer, Compression.NONE, ConsistencyLevel.QUORUM);
```

Comment:
Cassandra 1.2 has some support for `execute_cql3_query`. We already ran into this when writing `CqlExecutor`.